src/lib.rs

```diff
@@ -18,104 +18,79 @@
 //! [`DelayQueue`]: struct.DelayQueue.html
 //! [`Interval`]: struct.Interval.html
 
-use futures::stream::poll_fn;
-use futures::{try_ready, Async, Stream};
-use mio::unix::EventedFd;
-use mio::{Evented, Poll, PollOpt, Ready, Token};
-use std::io::Result;
-use std::os::unix::io::AsRawFd;
+use std::io::{Error, Result};
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::pin::Pin;
+use std::task::{Context, Poll};
 use std::time::{Duration, Instant};
+
+use futures::ready;
 use timerfd::{SetTimeFlags, TimerFd as InnerTimerFd, TimerState};
-use tokio_reactor::PollEvented;
+use tokio::io::unix::AsyncFd;
+use tokio::io::{AsyncRead, Interest, ReadBuf};
 
 pub use timerfd::ClockId;
 
 mod delay;
+/*
 mod delay_queue;
+*/
 mod interval;
 
 pub use delay::Delay;
-pub use delay_queue::DelayQueue;
 pub use interval::Interval;
+/*
+pub use delay_queue::DelayQueue;
+*/
 
-struct Inner(InnerTimerFd);
-
-impl Evented for Inner {
-    fn register(&self, poll: &Poll, token: Token, interest: Ready, opts: PollOpt) -> Result<()> {
-        poll.register(&EventedFd(&self.0.as_raw_fd()), token, interest, opts)
-    }
-
-    fn reregister(&self, poll: &Poll, token: Token, interest: Ready, opts: PollOpt) -> Result<()> {
-        poll.reregister(&EventedFd(&self.0.as_raw_fd()), token, interest, opts)
-    }
-
-    fn deregister(&self, poll: &Poll) -> Result<()> {
-        poll.deregister(&EventedFd(&self.0.as_raw_fd()))
-    }
-}
-
-pub struct TimerFd(PollEvented<Inner>);
+pub struct TimerFd(AsyncFd<InnerTimerFd>);
 
 impl TimerFd {
     pub fn new(clock: ClockId) -> std::io::Result<Self> {
-        let inner = PollEvented::new(Inner(InnerTimerFd::new_custom(clock, true, true)?));
+        let fd = InnerTimerFd::new_custom(clock, true, true)?;
+        let inner = AsyncFd::with_interest(fd, Interest::READABLE)?;
         Ok(TimerFd(inner))
     }
 
     fn set_state(&mut self, state: TimerState, flags: SetTimeFlags) {
-        (self.0).get_mut().0.set_state(state, flags);
+        (self.0).get_mut().set_state(state, flags);
     }
+}
 
-    fn poll_read(&mut self) -> Result<Async<()>> {
-        let ready = try_ready!(self.0.poll_read_ready(Ready::readable()));
-        self.0.get_mut().0.read();
-        self.0.clear_read_ready(ready)?;
-        Ok(Async::Ready(()))
+fn read_u64(fd: RawFd) -> Result<u64> {
```

Comment:
Arguably returning an `[u8; 8]` here makes more sense if you're just going to turn it back into an array afterwards?

On the other hand, maybe the semantics of a timer fd are such that an u64 is what makes sense?

Reply:
Well, semantically u64 makes more sense (see my other comment), but it's not exposed anywhere, so an array would probably be better.